plugins/apps/threatbus_zeek/plugin.py

```diff
@@ -11,129 +12,180 @@
     map_to_internal,
     map_management_message,
 )
-import time
+from typing import Callable, Dict, List, Union
 
 """Zeek network monitor - plugin for Threat Bus"""
 
 plugin_name = "zeek"
 lock = threading.Lock()
-subscriptions = dict()
-
-
-def validate_config(config):
-    assert config, "config must not be None"
-    config["host"].get(str)
-    config["port"].get(int)
-    config["module_namespace"].get(str)
-
-
-def rand_string(length):
-    """Generates a pseudo-random string with the requested length"""
-    letters = string.ascii_lowercase
-    return "".join(random.choice(letters) for i in range(length))
-
-
-def manage_subscription(
-    ep, module_namespace, task, subscribe_callback, unsubscribe_callback
-):
-    global lock, subscriptions
-    rand_suffix_length = 10
-    if type(task) is Subscription:
-        # point-to-point topic and queue for that particular subscription
-        logger.info(f"Received subscription for topic: {task.topic}")
-        p2p_topic = task.topic + rand_string(rand_suffix_length)
-        p2p_q = Queue()
-        ack = broker.zeek.Event(
-            f"{module_namespace}::subscription_acknowledged", p2p_topic
-        )
-        ep.publish("threatbus/manage", ack)
-        subscribe_callback(task.topic, p2p_q, task.snapshot)
-        lock.acquire()
-        subscriptions[p2p_topic] = p2p_q
-        lock.release()
-    elif type(task) is Unsubscription:
-        logger.info(f"Received unsubscription from topic: {task.topic}")
-        threatbus_topic = task.topic[: len(task.topic) - rand_suffix_length]
-        p2p_q = subscriptions.get(task.topic, None)
-        if p2p_q:
-            unsubscribe_callback(threatbus_topic, p2p_q)
+subscriptions: Dict[str, JoinableQueue] = dict()  # p2p_topic => queue
+workers: List[threatbus.StoppableWorker] = list()
+
+
+class SubscriptionManager(threatbus.StoppableWorker):
+    def __init__(
+        self,
+        module_namespace: str,
+        ep: broker.Endpoint,
+        subscribe_callback: Callable,
+        unsubscribe_callback: Callable,
+    ):
+        """
+        @param module_namespace A Zeek namespace to accept events from
+        @param ep The broker endpoint used for listening
+        @param subscribe_callback The callback to invoke for new subscriptions
+        @param unsubscribe_callback The callback to invoke for revoked subscriptions
+        """
+        super(SubscriptionManager, self).__init__()
+        self.ep = ep
+        self.module_namespace = module_namespace
+        self.subscribe_callback = subscribe_callback
+        self.unsubscribe_callback = unsubscribe_callback
+        self.rand_suffix_length = 10
```

Comment:
Here it would be also nice to know: where does 10 come from? Experimentation?

Reply:
No strong opinion on that one. We used it before and tbh, I don't expect collisions because folks usually won't add thousands of Zeek instances.